README.rst
#############################################################
python-escpos - Python library to manipulate ESC/POS Printers
#############################################################

.. image:: https://travis-ci.org/python-escpos/python-escpos.svg?branch=master
    :target: https://travis-ci.org/python-escpos/python-escpos
    :alt: Continous Integration

.. image:: https://codecov.io/github/python-escpos/python-escpos/coverage.svg?branch=master
    :target: https://codecov.io/github/python-escpos/python-escpos?branch=master
    :alt: Code Coverage

.. image:: https://readthedocs.org/projects/python-escpos/badge/?version=latest
    :target: https://python-escpos.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status


Description
-----------

Python ESC/POS is a library which lets the user have access to all those printers handled
by ESC/POS commands, as defined by Epson, from a Python application.

The library tries to implement the functions provided by the ESC/POS-commandset and supports sending text, images,
barcodes and qr-codes to the printer.

Text can be aligned/justified and fonts can be changed by size, type and weight.

Also, this module handles some hardware functionalities like cutting paper, control characters, printer reset
and similar functions.

Since supported commands differ from printer to printer the software tries to automatically apply the right
settings for the printer that you set. These settings are handled by
`escpos-printer-db <https://github.com/receipt-print-hq/escpos-printer-db>`_ which is also used in
`escpos-php <https://github.com/mike42/escpos-php>`_.

Dependencies
------------

This library makes use of:

* `pyusb <https://github.com/walac/pyusb>`_ for USB-printers
* `Pillow <https://github.com/python-pillow/Pillow>`_ for image printing
* `qrcode <https://github.com/lincolnloop/python-qrcode>`_ for the generation of QR-codes
* `pyserial <https://github.com/pyserial/pyserial>`_ for serial printers
* `python-barcode <https://github.com/WhyNotHugo/python-barcode>`_ for the generation of barcodes

Documentation and Usage
-----------------------

The basic usage is:

.. code:: python

    from escpos.printer import Usb

    """ Seiko Epson Corp. Receipt Printer (EPSON TM-T88III) """
    p = Usb(0x04b8, 0x0202, 0, profile="TM-T88III")
    p.text("Hello World\n")
    p.image("logo.gif")
    p.barcode('1324354657687', 'EAN13', 64, 2, '', '')
    p.cut()


Another example based on the Network printer class:

.. code:: python

    from escpos.printer import Network

    kitchen = Network("192.168.1.100") #Printer IP Address
    kitchen.text("Hello World\n")
    kitchen.barcode('1324354657687', 'EAN13', 64, 2, '', '')
    kitchen.cut()

Another example based on the Serial printer class:

.. code:: python

    from escpos.printer import Serial

    """ 9600 Baud, 8N1, Flow Control Enabled """
    p = Serial(devfile='/dev/tty.usbserial',
               baudrate=9600,
               bytesize=8,
               parity='N',
               stopbits=1,
               timeout=1.00,
               dsrdtr=True)

    p.text("Hello World\n")
    p.qr("You can readme from your smartphone")
    p.cut()


The full project-documentation is available on `Read the Docs <https://python-escpos.readthedocs.io>`_.

Contributing
------------

This project is open for any contribution! Please see `CONTRIBUTING.rst <https://python-escpos.readthedocs.io/en/latest/dev/contributing.html>`_ for more information.


Disclaimer
----------

None of the vendors cited in this project agree or endorse any of the patterns or implementations.
Its names are used only to maintain context.


CHANGELOG.rst
*********
Changelog
*********
2020-05-12 - Version 3.0a8 - "Only Slightly Bent"
-------------------------------------------------
This release is the ninth alpha release of the new version 3.0.
Please be aware that the API is subject to change until v3.0 is
released.

This release drops support for Python 2, requiring at least
version 3.5 of Python.

changes
^^^^^^^
- Drop support for Python 2 and mark in setuptools as only supporting 3.5 and upwards
- remove landscape.io badge
- replace viivakoodi with python-barcode which is maintained
- add configuration for Visual Studio Code
- use pkg_resources for the retrieval of the capabilities.json

contributors
^^^^^^^^^^^^
- Romain Porte
- Patrick Kanzler

2020-05-09 - Version 3.0a7 - "No Fixed Abode"
---------------------------------------------
This release is the eight alpha release of the new version 3.0.
Please be aware that the API is subject to change until v3.0
is released.

This release also marks the point at which the project transitioned
to having only a master-branch (and not an additional development branch).

changes
^^^^^^^
- add Exception for NotImplementedError in detach_kernel_driver
- update installation information
- update and improve documentation
- add error handling to image centering flag
- update and fix tox and CI environment, preparing drop of support for Python 2

contributors
^^^^^^^^^^^^
- Alexander Bougakov
- Brian
- Yaisel Hurtado
- Maximilan Wagenbach
- Patrick Kanzler

2019-06-19 - Version 3.0a6 - "Mistake not..."
---------------------------------------------
This release is the seventh alpha release of the new version 3.0.
Please be aware that the API is subject to change until v3.0 is
released.

changes
^^^^^^^
- fix inclusion of the capabilities-file
- execute CI jobs also on Windows and macOS-targets
- improve documentation

contributors
^^^^^^^^^^^^
- Patrick Kanzler

2019-06-16 - Version 3.0a5 - "Lightly Seared On The Reality Grill"
------------------------------------------------------------------
This release is the sixth alpha release of the new version 3.0. Please
be aware that the API is subject to change until v3.0 is released.

changes
^^^^^^^
- allow arbitrary USB arguments in USB-class
- add Win32Raw-Printer on Windows-platforms
- add and improve Windows support of USB-class
- use pyyaml safe_load()
- improve doc
- implement _read method of Network printer class

contributors
^^^^^^^^^^^^
- Patrick Kanzler
- Gerard Marull-Paretas
- Ramon Poca
- akeonly
- Omer Akram
- Justin Vieira

2018-05-15 - Version 3.0a4 - "Kakistocrat"
------------------------------------------
This release is the fifth alpha release of the new version 3.0. Please
be aware that the API will still change until v3.0 is released.

changes
^^^^^^^
- raise exception when TypeError occurs in cashdraw (#268)
- Feature/clear content in dummy printer (#271)
- fix is_online() (#282)
- improve documentation
- Modified submodule to always pull from master branch (#283)
- parameter for implementation of nonnative qrcode (#289)
- improve platform independence (#296)

contributors
^^^^^^^^^^^^
- Christoph Heuel
- Patrick Kanzler
- kennedy
- primax79
- reck31
- Thijs Triemstra

2017-10-08 - Version 3.0a3 - "Just Testing"
-------------------------------------------
This release is the fourth alpha release of the new version 3.0. Please
be aware that the API will still change until v3.0 is released.

changes
^^^^^^^
- minor changes in documentation, tests and examples
- pickle capabilities for faster startup
- first implementation of centering images and QR
- check barcodes based on regex

contributors
^^^^^^^^^^^^
- Patrick Kanzler
- Lucy Linder
- Romain Porte
- Sergio Pulgarin

2017-08-04 - Version 3.0a2 - "It's My Party And I'll Sing If I Want To"
-----------------------------------------------------------------------
This release is the third alpha release of the new version 3.0. Please
be aware that the API will still change until v3.0 is released.

changes
^^^^^^^
- refactor of the set-method
- preliminary support of POS "line display" printing
- improvement of tests
- added ImageWidthError
- list authors in repository
- add support for software-based barcode-rendering
- fix SerialException when trying to close device on __del__
- added the DLE EOT querying command for USB and Serial
- ensure QR codes have a large enough border
- make feed for cut optional
- fix the behavior of horizontal tabs
- added test script for hard an soft barcodes
- implemented paper sensor querying command
- added weather forecast example script
- added a method for simpler newlines

contributors
^^^^^^^^^^^^
- csoft2k
- Patrick Kanzler
- mrwunderbar666
- Romain Porte
- Ahmed Tahri

2017-03-29 - Version 3.0a1 - "Headcrash"
----------------------------------------
This release is the second alpha release of the new version 3.0. Please
be aware that the API will still change until v3.0 is released.

changes
^^^^^^^
- automatically upload releases to GitHub
- add environment variable ESCPOS_CAPABILITIES_FILE
- automatically handle cases where full cut or partial cut is not available
- add print_and_feed

contributors
^^^^^^^^^^^^
- Sam Cheng
- Patrick Kanzler
- Dmytro Katyukha

2017-01-31 - Version 3.0a - "Grey Area"
---------------------------------------
This release is the first alpha release of the new version 3.0. Please
be aware that the API will still change until v3.0 is released.

changes
^^^^^^^
- change the project's license to MIT in accordance with the contributors (see python-escpos/python-escpos#171)
- feature: add "capabilities" which are shared with escpos-php, capabilities are stored in
  `escpos-printer-db <https://github.com/receipt-print-hq/escpos-printer-db>`_

src/escpos/cli.py
#!/usr/bin/env python
# PYTHON_ARGCOMPLETE_OK
""" CLI

This module acts as a command line interface for python-escpos. It mirrors
closely the available ESCPOS commands while adding a couple extra ones for convenience.

It requires you to have a configuration file. See documentation for details.

"""


import argparse

try:
    import argcomplete
except ImportError:
    # this CLI works nevertheless without argcomplete
    pass  # noqa
import sys
import six
from . import config
from . import version


# Must be defined before it's used in DEMO_FUNCTIONS
def str_to_bool(string):
    """Used as a type in argparse so that we get back a proper
    bool instead of always True
    """
    return string.lower() in ("y", "yes", "1", "true")


# A list of functions that work better with a newline to be sent after them.
REQUIRES_NEWLINE = ("qr", "barcode", "text", "block_text")


# Used in demo method
# Key: The name of escpos function and the argument passed on the CLI. Some
#   manual translation is done in the case of barcodes_a -> barcode.
# Value: A list of dictionaries to pass to the escpos function as arguments.
DEMO_FUNCTIONS = {
    "text": [
        {
            "txt": "Hello, World!\n",
        }
    ],
    "qr": [
        {"content": "This tests a QR code"},
        {"content": "https://en.wikipedia.org/"},
    ],
    "barcodes_a": [
        {"bc": "UPC-A", "code": "13243546576"},
        {"bc": "UPC-E", "code": "132435"},
        {"bc": "EAN13", "code": "1324354657687"},
        {"bc": "EAN8", "code": "1324354"},
        {"bc": "CODE39", "code": "TEST"},
        {"bc": "ITF", "code": "55867492279103"},
        {"bc": "NW7", "code": "A00000000A"},
    ],
    "barcodes_b": [
        {"bc": "UPC-A", "code": "13243546576", "function_type": "B"},
        {"bc": "UPC-E", "code": "132435", "function_type": "B"},
        {"bc": "EAN13", "code": "1324354657687", "function_type": "B"},
        {"bc": "EAN8", "code": "1324354", "function_type": "B"},
        {"bc": "CODE39", "code": "TEST", "function_type": "B"},
        {"bc": "ITF", "code": "55867492279103", "function_type": "B"},
        {"bc": "NW7", "code": "A00000000A", "function_type": "B"},
        {"bc": "CODE93", "code": "A00000000A", "function_type": "B"},
        {"bc": "CODE93", "code": "1324354657687", "function_type": "B"},
        {"bc": "CODE128A", "code": "TEST", "function_type": "B"},
        {"bc": "CODE128B", "code": "TEST", "function_type": "B"},
        {"bc": "CODE128C", "code": "TEST", "function_type": "B"},
        {"bc": "GS1-128", "code": "00123456780000000001", "function_type": "B"},
        {
            "bc": "GS1 DataBar Omnidirectional",
            "code": "0000000000000",
            "function_type": "B",
        },
        {"bc": "GS1 DataBar Truncated", "code": "0000000000000", "function_type": "B"},
        {"bc": "GS1 DataBar Limited", "code": "0000000000000", "function_type": "B"},
        {"bc": "GS1 DataBar Expanded", "code": "00AAAAAAA", "function_type": "B"},
    ],
}

# Used to build the CLI
# A list of dictionaries. Each dict is a CLI argument.
# Keys:
# parser: A dict of args for command_parsers.add_parser
# defaults: A dict of args for subparser.set_defaults
# arguments: A list of dicts of args for subparser.add_argument
ESCPOS_COMMANDS = [
    {
        "parser": {
            "name": "qr",
            "help": "Print a QR code",
        },
        "defaults": {
            "func": "qr",
        },
        "arguments": [
            {
                "option_strings": ("--content",),
                "help": "Text to print as a qr code",
                "required": True,
            },
            {
                "option_strings": ("--size",),
                "help": "QR code size (1-16) [default:3]",
                "required": False,
                "type": int,
            },
        ],
    },
    {
        "parser": {
            "name": "barcode",
            "help": "Print a barcode",
        },
        "defaults": {
            "func": "barcode",
        },
        "arguments": [
            {
                "option_strings": ("--code",),
                "help": "Barcode data to print",
                "required": True,
            },
            {
                "option_strings": ("--bc",),
                "help": "Barcode format",
                "required": True,
            },
            {
                "option_strings": ("--height",),
                "help": "Barcode height in px",
                "type": int,
            },
            {
                "option_strings": ("--width",),
                "help": "Barcode width",
                "type": int,
            },
            {
                "option_strings": ("--pos",),
                "help": "Label position",
                "choices": ["BELOW", "ABOVE", "BOTH", "OFF"],
            },
            {
                "option_strings": ("--font",),
                "help": "Label font",
                "choices": ["A", "B"],
            },
            {
                "option_strings": ("--align_ct",),
                "help": "Align barcode center",
                "type": str_to_bool,
            },
            {
                "option_strings": ("--function_type",),
                "help": "ESCPOS function type",
                "choices": ["A", "B"],
            },
        ],
    },
    {
        "parser": {
            "name": "text",
            "help": "Print plain text",
        },
        "defaults": {
            "func": "text",
        },
        "arguments": [
            {
                "option_strings": ("--txt",),
                "help": "Plain text to print",
                "required": True,
            }
        ],
    },
    {
        "parser": {
            "name": "block_text",
            "help": "Print wrapped text",
        },
        "defaults": {
            "func": "block_text",
        },
        "arguments": [
            {
                "option_strings": ("--txt",),
                "help": "block_text to print",
                "required": True,
            },
            {
                "option_strings": ("--columns",),
                "help": "Number of columns",
                "type": int,
            },
        ],
    },
    {
        "parser": {
            "name": "cut",
            "help": "Cut the paper",
        },
        "defaults": {
            "func": "cut",
        },
        "arguments": [
            {
                "option_strings": ("--mode",),
                "help": "Type of cut",
                "choices": ["FULL", "PART"],
            },
        ],
    },
    {
        "parser": {
            "name": "cashdraw",
            "help": "Kick the cash drawer",
        },
        "defaults": {
            "func": "cashdraw",
        },
        "arguments": [
            {
                "option_strings": ("--pin",),
                "help": "Which PIN to kick",
                "choices": [2, 5],
            },
        ],
    },
    {
        "parser": {
            "name": "image",
            "help": "Print an image",
        },
        "defaults": {
            "func": "image",
        },
        "arguments": [
            {
                "option_strings": ("--img_source",),
                "help": "Path to image",
                "required": True,
            },
            {
                "option_strings": ("--impl",),
                "help": "Implementation to use",
                "choices": ["bitImageRaster", "bitImageColumn", "graphics"],
            },
            {
                "option_strings": ("--high_density_horizontal",),
                "help": "Image density (horizontal)",
                "type": str_to_bool,
            },
            {
                "option_strings": ("--high_density_vertical",),
                "help": "Image density (vertical)",
                "type": str_to_bool,
            },
        ],
    },
    {
        "parser": {
            "name": "fullimage",
            "help": "Print a fullimage",
        },
        "defaults": {
            "func": "fullimage",
        },
        "arguments": [
            {
                "option_strings": ("--img",),
                "help": "Path to img",
                "required": True,
            },
            {
                "option_strings": ("--max_height",),
                "help": "Max height of image in px",
                "type": int,
            },
            {
                "option_strings": ("--width",),
                "help": "Max width of image in px",
                "type": int,
            },
            {
                "option_strings": ("--histeq",),
                "help": "Equalize the histrogram",
                "type": str_to_bool,
            },
            {
                "option_strings": ("--bandsize",),
                "help": "Size of bands to divide into when printing",
                "type": int,
            },
        ],
    },
    {
        "parser": {
            "name": "charcode",
            "help": "Set character code table",
        },
        "defaults": {
            "func": "charcode",
        },
        "arguments": [
            {
                "option_strings": ("--code",),
                "help": "Character code",
                "required": True,
            },
        ],
    },
    {
        "parser": {
            "name": "set",
            "help": "Set text properties",
        },
        "defaults": {
            "func": "set",
        },
        "arguments": [
            {
                "option_strings": ("--align",),
                "help": "Horizontal alignment",
                "choices": ["left", "center", "right"],
            },
            {
                "option_strings": ("--font",),
                "help": "Font choice",
                "choices": ["left", "center", "right"],
            },
            {
                "option_strings": ("--text_type",),
                "help": "Text properties",
                "choices": ["B", "U", "U2", "BU", "BU2", "NORMAL"],
            },
            {
                "option_strings": ("--width",),
                "help": "Width multiplier",
                "type": int,
            },
            {
                "option_strings": ("--height",),
                "help": "Height multiplier",
                "type": int,
            },
            {
                "option_strings": ("--density",),
                "help": "Print density",
                "type": int,
            },
            {
                "option_strings": ("--invert",),
                "help": "White on black printing",
                "type": str_to_bool,
            },
            {
                "option_strings": ("--smooth",),
                "help": "Text smoothing. Effective on >:  4x4 text",
                "type": str_to_bool,
            },
            {
                "option_strings": ("--flip",),
                "help": "Text smoothing. Effective on >:  4x4 text",
                "type": str_to_bool,
            },
        ],
    },
    {
        "parser": {
            "name": "hw",
            "help": "Hardware operations",
        },
        "defaults": {
            "func": "hw",
        },
        "arguments": [
            {
                "option_strings": ("--hw",),
                "help": "Operation",
                "choices": ["INIT", "SELECT", "RESET"],
                "required": True,
            },
        ],
    },
    {
        "parser": {
            "name": "control",
            "help": "Control sequences",
        },
        "defaults": {
            "func": "control",
        },
        "arguments": [
            {
                "option_strings": ("--ctl",),
                "help": "Control sequence",
                "choices": ["LF", "FF", "CR", "HT", "VT"],
                "required": True,
            },
            {
                "option_strings": ("--pos",),
                "help": "Horizontal tab position (1-4)",
                "type": int,
            },
        ],
    },
    {
        "parser": {
            "name": "panel_buttons",
            "help": "Controls panel buttons",
        },
        "defaults": {
            "func": "panel_buttons",
        },
        "arguments": [
            {
                "option_strings": ("--enable",),
                "help": "Feed button enabled",
                "type": str_to_bool,
                "required": True,
            },
        ],
    },
    {
        "parser": {
            "name": "raw",
            "help": "Raw data",
        },
        "defaults": {
            "func": "_raw",
        },
        "arguments": [
            {
                "option_strings": ("--msg",),
                "help": "Raw data to send",
                "required": True,
            },
        ],
    },
]


def main():
    """

    Handles loading of configuration and creating and processing of command
    line arguments. Called when run from a CLI.

    """

    parser = argparse.ArgumentParser(
        description="CLI for python-escpos",
        epilog="Printer configuration is defined in the python-escpos config"
        "file. See documentation for details.",
    )

    parser.register("type", "bool", str_to_bool)

    # Allow config file location to be passed
    parser.add_argument(
        "-c",
        "--config",
        help="Alternate path to the configuration file",
    )

    # Everything interesting runs off of a subparser so we can use the format
    # cli [subparser] -args
    command_subparsers = parser.add_subparsers(
        title="ESCPOS Command",
        dest="parser",
    )
    # fix inconsistencies in the behaviour of some versions of argparse
    command_subparsers.required = False  # force 'required' testing

    # Build the ESCPOS command arguments
    for command in ESCPOS_COMMANDS:
        parser_command = command_subparsers.add_parser(**command["parser"])
        parser_command.set_defaults(**command["defaults"])
        for argument in command["arguments"]:
            option_strings = argument.pop("option_strings")
            parser_command.add_argument(*option_strings, **argument)

    # Build any custom arguments
    parser_command_demo = command_subparsers.add_parser(
        "demo", help="Demonstrates various functions"
    )
    parser_command_demo.set_defaults(func="demo")
    demo_group = parser_command_demo.add_mutually_exclusive_group()
    demo_group.add_argument(
        "--barcodes-a",
        help="Print demo barcodes for function type A",
        action="store_true",
    )
    demo_group.add_argument(
        "--barcodes-b",
        help="Print demo barcodes for function type B",
        action="store_true",
    )
    demo_group.add_argument(
        "--qr",
        help="Print some demo QR codes",
        action="store_true",
    )
    demo_group.add_argument(
        "--text",
        help="Print some demo text",
        action="store_true",
    )

    parser_command_version = command_subparsers.add_parser(
        "version", help="Print the version of python-escpos"
    )
    parser_command_version.set_defaults(version=True)

    # hook in argcomplete
    if "argcomplete" in globals():
        argcomplete.autocomplete(parser)

    # Get only arguments actually passed
    args_dict = vars(parser.parse_args())
    if not args_dict:
        parser.print_help()
        sys.exit()
    command_arguments = dict(
        [k, v] for k, v in six.iteritems(args_dict) if v is not None
    )

    # If version should be printed, do this, then exit
    print_version = command_arguments.pop("version", None)
    if print_version:
        print(version.version)
        sys.exit()

    # If there was a config path passed, grab it
    config_path = command_arguments.pop("config", None)

    # Load the configuration and defined printer
    saved_config = config.Config()
    saved_config.load(config_path)
    printer = saved_config.printer()

    if not printer:
        raise Exception("No printers loaded from config")

    target_command = command_arguments.pop("func")

    # remove helper-argument 'parser' from dict
    command_arguments.pop("parser", None)

    if hasattr(printer, target_command):
        # print command with args
        getattr(printer, target_command)(**command_arguments)
        if target_command in REQUIRES_NEWLINE:
            printer.text("\n")
    else:
        command_arguments["printer"] = printer
        globals()[target_command](**command_arguments)


def demo(printer, **kwargs):
    """
    Prints demos. Called when CLI is passed `demo`. This function
    uses the DEMO_FUNCTIONS dictionary.

    :param printer: A printer from escpos.printer
    :param kwargs: A dict with a key for each function you want to test. It's
        in this format since it usually comes from argparse.
    """
    for demo_choice in kwargs.keys():
        command = getattr(
            printer,
            demo_choice.replace("barcodes_a", "barcode").replace(
                "barcodes_b", "barcode"
            ),
        )
        for params in DEMO_FUNCTIONS[demo_choice]:
            command(**params)
        printer.cut()


if __name__ == "__main__":
    main()
